product_main_supplier/models/product_template.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2020 Coop IT Easy SCRL fs
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from datetime import date
+
+from odoo import api, fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    # fixme could product.main_seller_id be used instead? it seems that
+    # “seller” and “supplier” are used interchangeably in this
+    # class. is this on purpose?
+
+    # "top" is different from "main":
+    # top_supplierinfo_id is a technical field,
+    # and it's not the main supplier
+    # that is selected but the first one.
+    # So top_supplier reflect more the reality.
+    top_supplierinfo_id = fields.Many2one(
+        comodel_name="product.supplierinfo",
+        compute="_compute_top_supplierinfo_id",
+        store=True,
+    )
+    main_seller_id = fields.Many2one(
+        string="Main Seller",
+        comodel_name="res.partner",
+        related="top_supplierinfo_id.name",
+    )
+    main_seller_id_product_code = fields.Char(
+        string="Main Seller Product Code",
+        related="top_supplierinfo_id.product_code",
+    )
+
+    @api.multi
+    @api.depends("seller_ids", "seller_ids.date_start")
+    def _compute_top_supplierinfo_id(self):
+        for product in self:
+            # Compute the associated seller that
+            # - has the most recent start date
+            # - start_date is before today
+            sellers_ids = product._get_top_supplier_info()
+            product.top_supplierinfo_id = sellers_ids and sellers_ids[0] or False
+
+    def _get_top_supplier_info(self):
+        # fixme this function either returns a supplier or a collection.
+        #  wouldn’t it be more logical to return a supplier or None?
+
+        # supplierinfo w/o date_start come first
+        def sort_date_first(seller):
+            if seller.date_start:
+                return seller.date_start
+            else:
+                return date.max
+
+        suppliers = self.seller_ids.sorted(key=sort_date_first, reverse=True)
+        if suppliers:
+            return suppliers[0]
+        else:
+            return suppliers
```

Comment:
@victor-champonnois Take a look at this : #459 I also have added tests.

And the `start_date is before today` seams to be false to me. It just sort supplier by `date_start desc`.